bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java

```diff
@@ -77,27 +76,34 @@ public void processBindingConfiguration(String context, Item item, String bindin
 			throw new BindingConfigParseException("context is not permitted to be null for item "+item);
 		}
 
-		Set<Item> items = contextMap.get(context);
-		if (items==null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if (items == null) {
+				items = new HashSet<Item>();
+				contextMap.put(context, items);
+			}
+
+			items.add(item);
 		}
-			
-		items.add(item);
 	}
 
 	/**
 	 * {@inheritDoc}
 	 */
 	public void removeConfigurations(String context) {
-		Set<Item> items = contextMap.get(context);
-		if(items!=null) {
-			for(Item item : items) {
+		Set<Item> items = null;
+		synchronized (contextMap) {
+			items = contextMap.get(context);
+			if (items != null) {
+				contextMap.remove(context);
+			}
+		}
+		if (items != null) {
+			for (Item item : items) {
 				// we remove all binding configurations for all items
 				bindingConfigs.remove(item.getName());
 				notifyListeners(item);
 			}
-			contextMap.remove(context);
 		}
 	}
 	
```

Comment:
It's threadsafe, but my concern about remove [contextMap] before notify [bindingConfig] needs to be validated.  For example, with the changes in place someone previously using the `contextMap` object, via some sort of call into these Core methods during the `notifyListeners` phase might get a surprise ;)  

I don't know if that can actually happen, since I haven't looked over this part in that much detail, so it's theoretical at this point.